nni/contribute/distillation/uid_dataset.py

```diff
@@ -0,0 +1,117 @@
+
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Type
+
+import torch
+from torch.utils.data import Dataset, IterableDataset
+
+
+class _UidDataset(Dataset):
+    def __init__(self, dataset: Dataset, *args, **kwargs):
+        assert not isinstance(dataset, IterableDataset), 'Do not support `IterableDataset`.'
+        self._dataset = dataset
+        self._replay_mode = False
+
+    def __len__(self):
+        return len(self._dataset)  # type: ignore
+
+    def __getitem__(self, index):
+        return None, self._dataset.__getitem__(index)
+
+    def get_origin_dataset(self):
+        return self._dataset
+
+    def observe(self):
+        """
+        Observe mode means this dataset is on processing by `Preprocessor`
+        """
+        self._replay_mode = False
+
+    def replay(self):
+        """
+        Replay mode means this dataset will replay the previous samples.
+        """
+        self._replay_mode = True
+
+
+class IndexedDataset(_UidDataset):
+    """
+    Return (index, origin_sample).
+    """
+    def __getitem__(self, index):
+        return str(index), self._dataset.__getitem__(index)
+
+
+class HashedDataset(_UidDataset):
+    """
+    Return (hash(origin_sample), origin_sample).
+    """
+    def __init__(self, dataset: Dataset, hash_fn: Callable[[Any], str]):
+        super().__init__(dataset)
+        self._hash_fn = hash_fn
+
+    def __getitem__(self, index):
+        sample = self._dataset.__getitem__(index)
+        return self._hash_fn(sample), sample
+
+
+class AugmentationDataset(_UidDataset):
+    """
+    Return (previous_uid/seed, origin_sample).
+    """
+    def __init__(self, dataset: Dataset, transform: Callable[[Any], Any], seed: int | None = None,
+                 aux_dataset_cls: Type[_UidDataset] | None = None, *aux_args, **aux_kwargs):
+        if isinstance(dataset, _UidDataset) and aux_dataset_cls is None:
+            _dataset = dataset
+        else:
+            aux_dataset_cls = IndexedDataset if aux_dataset_cls is None else aux_dataset_cls
+            assert issubclass(aux_dataset_cls, _UidDataset)
+            _dataset = aux_dataset_cls(dataset=dataset, *aux_args, **aux_kwargs)
+        super().__init__(_dataset)
+        self._transform = transform
+
+        self._rng = torch.Generator()
+        if seed is not None:
+            self._rng.manual_seed(seed=seed)
+
+        self._replay_mode = False
+        self._suid_seed = defaultdict(list)
+
+    def __getitem__(self, index):
+        suid, sample = self._dataset.__getitem__(index)
+
+        cpu_rng_state = torch.random.get_rng_state()
+
+        if self._replay_mode:
+            seed = self._suid_seed[str(suid)].pop(-1)
+            self._suid_seed[str(suid)].insert(0, seed)
+        else:
+            seed = self._generate_seed()
+            self._suid_seed[str(suid)].append(seed)
+        torch.random.manual_seed(seed)
+        sample = self._transform(sample)
```

Comment:
The function `self._transform` may use other random libraries, like `np.random` and `builtin.random`. We need to save and restore these states.

Reply:
I add `np.random`, `builtin.random`, `torch.random` rng save/load as default, do you think here need to give user an interface to customize save/load for some special random libraries we don't know?

Reply:
Great! Yes, it is better to provide an interface to customize the function `_default_manual_seed`.